torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -18,22 +18,40 @@ void confirmPendingUser(
   if (!futureMessage.hasError()) {
     auto rr = RemoteRet::fromMessage(futureMessage.constValue());
     TORCH_INTERNAL_ASSERT(rr->forkId() == expectedForkId);
+  } else {
+    // Handle errors, such as timeouts, by invoking the error handler on the
+    // rref.
+    auto rref_ptr = RRefContext::getInstance().getPendingUser(expectedForkId);
+    auto errorType = getRPCErrorType(futureMessage);
+    rref_ptr->handleError(errorType, futureMessage);
   }
   RRefContext::getInstance().delPendingUser(expectedForkId);
-  // Potentially propagate to the userRRef?
-  RRefContext::handleException(futureMessage);
 }
 
 c10::intrusive_ptr<RRef> finishCreatingOwnerRRef(
-    const FutureMessage& futureMessage) {
-  RRefContext::handleException(futureMessage);
-  auto rr = RemoteRet::fromMessage(futureMessage.constValue());
-  TORCH_INTERNAL_ASSERT(
-      rr->rrefId() == rr->forkId(),
-      "Expecting an OwnerRRef as RemoteRet but got a fork.");
-  auto& ctx = RRefContext::getInstance();
-  auto deletedRRef = ctx.delForkOfOwner(rr->rrefId(), rr->rrefId());
-  return deletedRRef;
+    const FutureMessage& futureMessage,
+    const RRefId& rrefId) {
+  if (futureMessage.hasError()) {
+    auto& ctx = RRefContext::getInstance();
+    // We expect to run this callback only after the OwnerRRef has been created,
+    // since this is only invoked when sending to self.
+    auto rref_ptr =
+        ctx.getOwnerRRef(rrefId, /* ensure created */ true)->constValue();
```

Comment:
This means if the RRef creation failed, it will crash one RPC thread?

Reply:
Does it make sense if we add a `getIfPresent/Exist` API? Or change the second arg to `blocking` instead of `forceCreated`?

Reply:
The only way this can crash is if the ownerRRef is somehow removed from the `owners_` map before we run this callback, and this callback is (currently) only run when creating an RRef on the self worker. In all codepaths, we call `addSelfAsFork` which puts the RRef into the `owners_` map. Although, possibly one case where this issue can happen is if we process a message causing us to run `delForkOfOwner`.

Reply:
I think one thing we could do is remove this forceCreated argument, and then call `wait`/`constValue` in `finishCreatingOwnerRRef`.

Reply:
Sounds good, will keep it the way it is.